crates/chain/src/local_chain.rs

```diff
@@ -0,0 +1,191 @@
+        let update = update.as_ref();
+        let update_tip = match update.keys().last().cloned() {
+            Some(tip) => tip,
+            None => return Ok(ChangeSet::default()),
+        };
+
+        // this is the latest height where both the update and local chain has the same block hash
+        let agreement_height = update
+            .iter()
+            .rev()
+            .find(|&(u_height, u_hash)| self.blocks.get(u_height) == Some(u_hash))
+            .map(|(&height, _)| height);
+
+        // the lower bound of the range to invalidate
+        let invalidate_lb = match agreement_height {
+            Some(height) if height == update_tip => u32::MAX,
+            Some(height) => height + 1,
+            None => 0,
+        };
+
+        // the first block's height to invalidate in the local chain
+        let invalidate_from_height = self.blocks.range(invalidate_lb..).next().map(|(&h, _)| h);
+
+        // the first block of height to invalidate (if any) should be represented in the update
+        if let Some(first_invalid_height) = invalidate_from_height {
+            if !update.contains_key(&first_invalid_height) {
+                return Err(UpdateNotConnectedError(first_invalid_height));
+            }
+        }
+
+        let mut changeset = BTreeMap::<u32, BlockHash>::new();
+        for (height, new_hash) in update {
+            let original_hash = self.blocks.get(height);
+            if Some(new_hash) != original_hash {
+                changeset.insert(*height, *new_hash);
+            }
+        }
+        Ok(ChangeSet(changeset))
+    }
+
+    /// Applies the given `changeset`.
+    pub fn apply_changeset(&mut self, mut changeset: ChangeSet) {
+        self.blocks.append(&mut changeset.0)
+    }
+
+    /// Updates [`LocalChain`] with an update [`LocalChain`].
+    ///
+    /// This is equivalent to calling [`determine_changeset`] and [`apply_changeset`] in sequence.
+    ///
+    /// [`determine_changeset`]: Self::determine_changeset
+    /// [`apply_changeset`]: Self::apply_changeset
+    pub fn apply_update(&mut self, update: Self) -> Result<ChangeSet, UpdateNotConnectedError> {
+        let changeset = self.determine_changeset(&update)?;
+        self.apply_changeset(changeset.clone());
+        Ok(changeset)
+    }
+
+    pub fn initial_changeset(&self) -> ChangeSet {
+        ChangeSet(self.blocks.clone())
+    }
+
+    pub fn heights(&self) -> BTreeSet<u32> {
+        self.blocks.keys().cloned().collect()
+    }
+}
+
+/// This is the return value of [`determine_changeset`] and represents changes to [`LocalChain`].
+///
+/// [`determine_changeset`]: LocalChain::determine_changeset
+#[derive(Debug, Default, Clone, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize, serde::Serialize),
+    serde(crate = "serde_crate")
+)]
+pub struct ChangeSet(pub(crate) BTreeMap<u32, BlockHash>);
+
+impl Deref for ChangeSet {
```

Comment:
Why can't we just use the BtreeMap itself. I don't think we need to wrap it here.